### DISCUSSION TITLE
Decryption of new FW

### DISCUSSION THREAD

**Discussion by qba667** [ludwig-v/wireless-carplay-dongle-reverse-engineering#155] [Ideas]
Hi @ludwig-v,
I have recently looked at your [STRACE](https://github.com/ludwig-v/wireless-carplay-dongle-reverse-engineering/blob/master/Reverse/STRACE.md) form recent FW. Will it be possible to attach a remote debugger to update operation?
It looks decrypted, unpacked ELF is stored in memory, so dumping memory will give us assembly used for decryption.
Some research is already made by the @Henkru:

https://github.com/Henkru/cplay2air-wifi-passphrase-patch/blob/main/README.md

@Henkru Maybe you can help with the topic? If I get unencrypted/unpacked ELF from https://github.com/ludwig-v/wireless-carplay-dongle-reverse-engineering/blob/master/Reverse/ARMimg_maker I can take a look inside.

**Comment by ludwig-v** [comment]
Hi ! Can you share how you extracted included binary ? You dumped memory ?

**Reply by qba667** [reply]
Hi @ludwig-v 
Following method shown by @Henkru. Compiled static linked gdbserver and and gdb (some patching is needed).
Started [ARMimg_maker](https://github.com/ludwig-v/wireless-carplay-dongle-reverse-engineering/blob/master/Reverse/ARMimg_maker) with GDB server. First extraction step extract code used to initialize AES - one of the method looks like key derivation method - a lot of iterations. 
Interesting fact that generated value seems to be used every time on AES operation. 
I have dump of this part but it is not very useful - parameters are not pointing to passphrase.
More interesting are parameters passed to ioctl - as this function is at the end doing decryption - pointers to buffers.
In second step firmware extractor is unpacked using the the AES implementation.
Update gzip is extracted using same method - I can see different values being passed.

At evening  I will attach few memory dumps.

If we want to made use of custom FW I would do it differently - patch extractor in runtime to not perform update but keep extracted update in RAM. Grab it, extract replace ARMimg_maker with old one, do custom FW changes, repack it using old method.

Next step will be to find what is exactly happing in kernel during ioctl call. Hope the computed byte array is the key and IV but the size is not matching (for AES 128 CBC) expecting 32 bytes (from IV size in assembly) but getting - 48 - possibly AES 192 CBC.

What do you think? For me the device is little bit limiting as kernel can not be replaced. I have searched a little about i.MX 6 Linux High Assurance Boot (HAB) there was a design flow that uses ASN1 parsing and let boot not signed kernel [CVE-2017-7932](https://community.nxp.com/docs/DOC-334996) but it looks current versions of MCUs are patched -I have dumped boot ROM of the MCU as well.
